src/log.h

```diff
@@ -0,0 +1,104 @@
+#pragma once
+
+#include <iostream>
+#include <sstream>
+
+#include "bpftrace.h"
+
+namespace bpftrace {
+
+// clang-format off
+enum class LogType
+{
+  debug,
+  info,
+  warning,
+  error
+};
+// clang-format on
+
+class Log
+{
+public:
+  static Log& get();
+  void take_input(LogType type, std::string&& input, std::ostream& output);
+  void take_input(LogType type,
+                  std::string&& input,
+                  const location& loc,
+                  std::ostream& output);
+  inline void set_source(std::string filename, std::string source)
+  {
+    src_ = source;
+    filename_ = filename;
+  }
+  inline const std::string& get_source()
+  {
+    return src_;
+  }
+  const std::string get_source_line(unsigned int n);
+
+  // Can only construct with get()
+  Log(const Log& other) = delete;
+  Log& operator=(const Log& other) = delete;
+
+private:
+  Log();
+  ~Log();
+  std::string src_;
+  std::string filename_;
+  void log_with_location(LogType,
+                         const location&,
+                         std::ostream&,
+                         const std::string&);
+};
+
+class LogStream
+{
+public:
+  LogStream(LogType type, std::ostream& out);
+  template <typename T>
+  LogStream& operator<<(const T& v)
+  {
+    buf_ << v;
+    return *this;
+  }
+  ~LogStream();
+
+private:
+  Log& sink_;
+  LogType type_;
+  std::ostream& out_;
+  std::ostringstream buf_;
+};
+
+class LogLocStream
+{
+public:
+  LogLocStream(LogType type, const location& loc, std::ostream& out);
+
+  template <typename T>
+  LogLocStream& operator<<(const T& v)
+  {
+    buf_ << v;
+    return *this;
+  }
+  ~LogLocStream();
+
+private:
+  Log& sink_;
+  LogType type_;
+  const location& loc_;
+  std::ostream& out_;
+  std::ostringstream buf_;
+};
+
+#define LOG(type, out) LogStream(type, out)
+// log with location
+#define LLOG(type, loc, out) LogLocStream(type, loc, out)
+// debugging log, use std::cerr by default
+#define DLOG                                                                   \
+  LogStream(LogType::debug, std::cerr)                                         \
+      << "[" << __FILE__ << ":" << __LINE__ << "]"
+#define DLOG_OUT(out)                                                          \
+  LogStream(LogType::debug, out) << "[" << __FILE__ << ":" << __LINE__ << "]"
```

Comment:
I'd prefer if we don't have `DLOG` and `DLOG_OUT`. In general it's good to keep API surface area small. Is there a strong reason we need these? I think it'd be better if LOG(LogType::debug) always printed line numbers anyways.